internal/controller/apiextensions/composite/reconciler.go

```diff
@@ -512,14 +512,17 @@ func (r *Reconciler) Reconcile(ctx context.Context, req reconcile.Request) (reco
 		return reconcile.Result{Requeue: true}, errors.Wrap(r.client.Status().Update(ctx, xr), errUpdateStatus)
 	}
 
+	orig := xr.GetCompositionReference()
 	if err := r.composite.SelectComposition(ctx, xr); err != nil {
 		log.Debug(errSelectComp, "error", err)
 		err = errors.Wrap(err, errSelectComp)
 		r.record.Event(xr, event.Warning(reasonResolve, err))
 		xr.SetConditions(xpv1.ReconcileError(err))
 		return reconcile.Result{Requeue: true}, errors.Wrap(r.client.Status().Update(ctx, xr), errUpdateStatus)
 	}
-	r.record.Event(xr, event.Normal(reasonResolve, "Successfully selected composition"))
+	if compRef := xr.GetCompositionReference(); (orig == nil) != (compRef == nil) || (orig != nil && *compRef != *orig) {
```

Comment:
Wouldn't this crash if `compRef` is `nil` but the `orig` is not? (which may not be possible depending on the implementation of SelectComposition though.

How about the following?

```suggestion
	if compRef := xr.GetCompositionReference(); compRef != nil && (orig == nil || *compRef != *orig) {
```

Reply:
It won't crash: if orig is also nil, `(orig == nil) != (compRef == nil)` is false, and the second only evaluated `orig != nil`. If orig is not nil, the first clause is true.

Reply:
What am I missing? https://go.dev/play/p/A-qNj9pnEs6

Reply:
Not the if clause is crashing, but the command afterwards.

Reply:
We don't need the case of unsetting of the reference. Have switched to your variant. It's easier. Thanks!